### ISSUE TITLE
Useless diagnostics from procmacro expansion

### ISSUE THREAD

**Issue by SomeoneToIgnore** [rust-lang/rust-analyzer#16514]
**rust-analyzer version**: rust-analyzer version: 0.4.1834-standalone (c48f14553 2024-02-07)
**rustc version**: rustc 1.75.0 (82e1608df 2023-12-21)

Open Zed, wait for r-a to load it and produce diagnostics in this file:

https://github.com/zed-industries/zed/blob/bde509fa749da60beb4ac378afd3ade15297bb91/crates/editor/src/display_map.rs#L1685
or
https://github.com/zed-industries/zed/blob/bde509fa749da60beb4ac378afd3ade15297bb91/crates/multi_buffer/src/multi_buffer.rs#L4764

<img width="1029" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/2690773/bbdf8fd0-a4ec-4c45-95c8-1e7cf10b5f40">

this diagnostic is very volatile, but usually appears around those files, sometimes producing dozens of errors, sometimes a couple.
Inside `#[gpui::test]` macro expansion there's indeed a `let mut ...` binding that does not need to be mutable, but
* rustc/cargo does not emit such warnings
* what the end procmacro user is supposed to do about it?

**Comment by SomeoneToIgnore** [comment]
Found a very minimalistic example:
```rust
thread_local! {
    static TEST: bool = false;
}
```
produces the `remove unnecessary else block` diagnostics that is totally useless

<img width="968" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/2690773/37b2b7ce-f0bb-404e-bdc2-59916610cab5">

**Comment by pacak** [comment]
FWIW the problem is not limited to proc macros. I'm getting an unhelpful diagnostic from `anyhow::bail!`. It's not like I can do anything about it.


`H     anyhow::bail!("no rmeta?");     ■ replace return <expr>; with <expr>`
